### PR TITLE
[sessiond] refactored get_charging_updates

### LINKED ISSUE #3211: TestGyCreditExhaustionRedirect failure

### PR DESCRIPTION
**Pull request by uri200**
Signed-off-by: Oriol Batalla <obatalla@fb.com>

## Summary

Refactored `get_charging_updates` to avoid fallthrough logic that may be confusing 

## Test Plan

make precommit_sm

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

**Review by themarwhal — APPROVED** [review]

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 1]
```diff
@@ -1385,6 +1385,12 @@ void SessionState::get_charging_updates(
         }
         grant->set_service_state(SERVICE_REDIRECTED, *credit_uc);
         action->set_redirect_server(grant->final_action_info.redirect_server);
+        // activate service
```
Comment (ymasmoudi) [inline lte/gateway/c/session_manager/SessionState.cpp:1388]: this comment seems unrelated. nit ?

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 2]
```diff
@@ -1395,31 +1401,47 @@ void SessionState::get_charging_updates(
         for (auto& rule : grant->final_action_info.restrict_rules) {
           restrict_rules->push_back(rule);
         }
+        // activate service
```
Comment (ymasmoudi) [inline lte/gateway/c/session_manager/SessionState.cpp:1404]: same as above.

File: lte/gateway/c/session_manager/SessionState.cpp [hunk 3]
```diff
@@ -1395,31 +1401,47 @@ void SessionState::get_charging_updates(
...
-        MLOG(MWARNING) << "Unexpected action type " << action_type << " for "
+        MLOG(MWARNING) << "Unexpected action type "
+                       << service_action_type_to_str(action_type) << " for "
                        << session_id_;
         break;
     }
   }
 }
 
+void SessionState::terminate_service_action(
```
Comment (ymasmoudi) [inline lte/gateway/c/session_manager/SessionState.cpp:1429]: this function is used by all actions (terminate/activate/restrict/redirect). 
worth renaming it to a more generic name.

**Review by ymasmoudi — APPROVED** [review]

File: lte/gateway/c/session_manager/SessionState.cpp (hunk 1 above)
Comment (uri200) [reply lte/gateway/c/session_manager/SessionState.cpp:1388]: `ACTIVATE_SERVICE` below is defined as `action->set_ambr(config_.get_apn_ambr());` 

Since with the refactor we lose the reference to this label I want to make sure that we know what this command is doing. Per its name it is not evident it is causing an activation so we better leave a comment for our future selves 

Maybe if you would, it would be better to define those steps with more detail like
``// activating redirected session``

``// terminating previous session``

(I am not even sure if this is the right flow)
